### ISSUE TITLE
How to flat the model about TFLite and remove the "while" loop?

### ISSUE THREAD

**Issue by panhu** [tensorflow/tensorflow#56734]
### 1. System information

- Linux:
- TensorFlow 2.9:

### 2. Code

The code of converter:

modelparh = r"model.h5"
model = tf.keras.models.load_model(modelparh, custom_objects = {'...'})

def representative_dataset_gen():
    for audio in x_train[0:500,:]:
        yield [audio.reshape(-1,x_train.shape[1],x_train.shape[2]).astype("float32")]

converter = TFLiteConverter.from_keras_model(model)

converter.optimizations = [tf.contrib.lite.Optimize.DEFAULT]

converter.representative_dataset = representative_dataset_gen

converter.target_spec.supported_ops=[tf.contrib.lite.OpsSet.TFLITE_BUILTINS,tf.contrib.lite.OpsSet.SELECT_TF_OPS,tf.contrib.lite.OpsSet.TFLITE_BUILTINS_INT8]

converter.inference_input_type = tf.int8
converter.inference_output_type = tf.int8
tflite_model = converter.convert()

savepath = r"model.tflite"
open(savepath, "wb").write(tflite_model)

-my issus is:
i can converter and quantification the model.but when i check the file of tflite by netron,find the model can't flat and have some "while" loop in gru graph. the graph like this:

while
1=0
2=0

**Comment by mohantym** [comment]
Hi @panhu!

Contrib has been removed in 2.x version. Could you just remove "Contrib" from tf.contrib.lite and let us know the difference. 
Please post a minimal stand alone code for the model to replicate the issue. 

Thank you!

**Comment by panhu** [comment]
Ok,The previous code was sent incorrectly，this is my simple code of convert:

modelparh = r"model.h5"
model = tf.keras.models.load_model(modelparh, custom_objects = {'...'})

converter = tf.lite.TFLiteConverter.from_keras_model(model)

converter.optimizations = [tf.lite.Optimize.DEFAULT]

converter.target_spec.supported_ops=[tf.lite.OpsSet.TFLITE_BUILTINS,tf.lite.OpsSet.SELECT_TF_OPS,tf.lite.OpsSet.TFLITE_BUILTINS_INT8]

tflite_model = converter.convert()

savepath = r"model.tflite"
open(savepath, "wb").write(tflite_model)

And,my model like this:

inp = Input(shape=(None, 84))

Layer_1 = Dense(48, activation='tanh')(inp)

gru_1 = GRU(48, activation='sigmoid', recurrent_activation='relu', return_sequences=True, kernel_regularizer=regularizers.l2(reg), recurrent_regularizer=regularizers.l2(reg))(Layer_1)

inp_2 = concatenate([inp, Layer_1, gru_1])

gru_2 = GRU(96, activation='tanh', recurrent_activation='relu', return_sequences=True, kernel_regularizer=regularizers.l2(reg), recurrent_regularizer=regularizers.l2(reg))(inp_2)

inp_3 = concatenate([gru_1, gru_2, inp])

gru_3 = GRU(192, activation='tanh', recurrent_activation='relu', return_sequences=True, 
kernel_regularizer=regularizers.l2(reg), recurrent_regularizer=regularizers.l2(reg))(inp_3)

out = Dense(44, activation='sigmoid', name='denoise_output', kernel_constraint=constraint, bias_constraint=constraint)(gru_3)

The same issus:

Checking the file of tflite by netron,find the model can't flat and have some "while" loop in gru graph. the loop of graph like this:

while
1=0
2=0

so,how to flat the model and remove the "while" loop

**Comment by mohantym** [comment]
Hi @sachinprasadhs !
Could you please look at this issue. Attached [gist](https://colab.sandbox.google.com/gist/mohantym/f1ffd1d15977681b86159c218bc2d4dd/git_56734.ipynb)  in 2.8 for reference. 
Thank you!

**Comment by sachinprasadhs** [comment]
@panhu, Could you please provide the screenshot of the graph in netron which you have mentioned above.

**Comment by panhu** [comment]
> @panhu, Could you please provide the screenshot of the graph in netron which you have mentioned above.
ok, i look the gru cell like this:
https://user-images.githubusercontent.com/11703018/178864665-8a75a585-8099-44f1-8291-e50ae6c9cc0a.PNG

**Comment by panhu** [comment]
More complete：
https://user-images.githubusercontent.com/11703018/178866754-305bfd0e-c778-4aac-a595-9377d22c381c.PNG

**Comment by sachinprasadhs** [comment]
Thanks for the detail. The TFlite converted model creates nodes in a optimized way, I don't think you can modify any of those.
Instead, you can follow any of the alternative way mentioned [here](https://www.tensorflow.org/lite/models/convert) to convert your model and see if that works for you. Thanks!

**Comment by jkaldal** [comment]
I'm having a similar problem ("while" loop) when converting a RNN network, with a custom RNN cell, to a tflite model. This also happens when using the LSTMcell instead of the custom cell or when using the GRUcell or GRU layer. However, this does NOT happen when using the LSTM layer.

```
timesteps = train_X.shape[1]
n_features = train_X.shape[2]
outputs = train_y.shape[1]

DROPOUT = 0.5
DENSE_UNITS = 40

# create model
model = tf.keras.Sequential([
    #tf.keras.layers.RNN(CustomRNNcell(100), input_shape=(timesteps, n_features)),
    tf.keras.layers.GRU(100, input_shape=(timesteps, n_features),),
    tf.keras.layers.Dropout(DROPOUT),
    tf.keras.layers.Dense(DENSE_UNITS, activation='relu'),
    tf.keras.layers.Dense(outputs, activation='softmax', name="output")
])

# create model for inference with batch_size = 1
model_inference = tf.keras.Sequential([
    #tf.keras.layers.RNN(CustomRNNcell(100), input_shape=(timesteps, n_features), batch_size= 1),
    tf.keras.layers.GRU(100, input_shape=(timesteps, n_features), batch_size=1),
    tf.keras.layers.Dense(DENSE_UNITS, batch_size= 1, activation='relu'),
    tf.keras.layers.Dense(outputs, batch_size= 1, activation='softmax', name="output")
])

#training paramenters
EPOCHS = 1
BATCH_SIZE = 64
LEARNING_RATE = 0.01

# compile model
opt = tf.keras.optimizers.Adam(learning_rate=LEARNING_RATE)
model.compile(optimizer=opt, loss='categorical_crossentropy', metrics=['accuracy'])

#train model
print("normal training..." )
history = model.fit(train_X, train_y, epochs=EPOCHS, validation_data=(validation_X, validation_y), shuffle=False,
    batch_size=BATCH_SIZE)

# move weights to inference
model_inference.set_weights(model.get_weights())
model_inference.compile(optimizer=opt, loss='categorical_crossentropy', metrics=['accuracy'])

# create converter
converter = tf.lite.TFLiteConverter.from_keras_model(model_inference)

# convert
tfmodel = converter.convert()
# save to tflite format
open(f'model.tflite', 'wb').write(tfmodel)
```

When I view my model in netron, this is what I get
![netron_gru_model](https://user-images.githubusercontent.com/70149133/200933132-fbb99d6f-c422-44cb-ab1f-f702547e6655.png)
This is what is returned from tf.lite.experimental.Analyzer.analyze(tfmodel)
```
Your TFLite model has '3' subgraph(s). In the subgraph description below,
T# represents the Tensor numbers. For example, in Subgraph#0, the RESHAPE op takes
tensor #0 and tensor #12 as input and produces tensor #13 as output.

Subgraph#0 main(T#0) -> [T#23]
  Op#0 RESHAPE(T#0, T#12) -> [T#13]
  Op#1 WHILE(T#3, T#3, T#1, T#4, T#4, T#13, Cond: Subgraph#1, Body: Subgraph#2) -> [T#14, T#15, T#16, T#17, T#18, T#19]
  Op#2 STRIDED_SLICE(T#16, T#9, T#10, T#11) -> [T#20]
  Op#3 FULLY_CONNECTED(T#20, T#7, T#5) -> [T#21]
  Op#4 FULLY_CONNECTED(T#21, T#8, T#6) -> [T#22]
  Op#5 SOFTMAX(T#22) -> [T#23]
. . .
```

However, when I create my model with an LSTM layer I get what I:
![netron_lstm_model](https://user-images.githubusercontent.com/70149133/200933083-d7881846-55e4-46d7-91ac-8374787766a9.png)
and from tf.lite.experimental.Analyzer.analyze(tfmodel)
```
Your TFLite model has '1' subgraph(s). In the subgraph description below,
T# represents the Tensor numbers. For example, in Subgraph#0, the UNIDIRECTIONAL_SEQUENCE_LSTM op takes
tensor #0 and tensor #18 and tensor #17 and tensor #16 and tensor #15 and tensor #10 and tensor #9 and tensor #8 and tensor #7 and tensor #-1 and tensor #-1 and tensor #-1 and tensor #14 and tensor #13 and tensor #12 and tensor #11 and tensor #-1 and tensor #-1 and tensor #4 and tensor #21 and tensor #-1 and tensor #-1 and tensor #-1 and tensor #-1 as input and produces tensor #22 as output.

Tensors of Subgraph#0
  T#0(serving_default_lstm_input:0) shape:[1, 128, 6], type:FLOAT32
  T#1(arith.constant) shape:[3], type:INT32 RO 12 bytes
  T#2(arith.constant1) shape:[3], type:INT32 RO 12 bytes
  T#3(arith.constant2) shape:[3], type:INT32 RO 12 bytes
  T#4(sequential_3/lstm/zeros) shape:[1, 100], type:FLOAT32
...
```

What can I do to remove the reshaping and while and get something similar to the second photo which was created by only changing
```
#tf.keras.layers.RNN(RNNcell(100), input_shape=(timesteps, n_features))
# or tf.keras.layers.RNN(tf.keras.layers.LSTMCell(100), input_shape=(timesteps, n_features))
# or tf.keras.layers.RNN(tf.keras.layers.GRUCell(100), input_shape=(timesteps, n_features))
# or
tf.keras.layers.GRU(100, input_shape=(timesteps, n_features)),
```
to
```
tf.keras.layers.LSTM(100, input_shape=(timesteps, n_features)),
```
